### PR TITLE
Update Joi to version 17

### PR DESCRIPTION
**Pull request by sprootshift**
Mostly building on @rmothilal work in https://github.com/tlivings/enjoi/pull/85.

`allOf` is supported through `schema.concat`
`types` were replaced with new Joi `extensions` syntax

### CODE REVIEW

**Comment by rmothilal** [comment]
@sprootshift thanks for this

**Comment by tlivings** [comment]
Thanks, reviewing...

File: lib/resolver.js [hunk 1]
```diff
@@ -1,39 +1,49 @@
...
+                coerce: {
+                    from: 'string',
+                    method(value) {
+                        if (typeof value !== 'string' || (value[0] !== '{' && !/^\s*\{/.test(value))) {
+                            return;
                         }
+
+                        try {
+                            return { value: Bourne.parse(value) };
+                        } catch (ignoreErr) { }
```
Comment (tlivings) [inline lib/resolver.js:26]: What is the impact of ignoring this error?

File: lib/resolver.js [hunk 2]
```diff
@@ -1,39 +1,49 @@
...
+                base: Joi.array(),
+                coerce: {
+                    from: 'string',
+                    method(value) {
+                        if (typeof value !== 'string' || (value[0] !== '[' && !/^\s*\[/.test(value))) {
+                            return;
+                        }
+                        try {
+                            return { value: Bourne.parse(value) };
+                        } catch (ignoreErr) { }
```
Comment (tlivings) [inline lib/resolver.js:41]: What is the impact of ignoring this error?

File: index.js [hunk 3]
```diff
@@ -8,7 +7,6 @@ const schemaSchema = Joi.alternatives(Joi.object().unknown(true), Joi.string()).
 
 const optionsSchema = Joi.object({
     subSchemas: Joi.object().unknown(true).allow(null),
-    types: Joi.object().unknown(true).allow(null),
```
Comment (tlivings) [inline index.js:11]: Note to self: breaking change.

File: lib/resolver.js [hunk 4]
```diff
@@ -1,39 +1,49 @@
-
 const Joi = require('@hapi/joi');
 const Util = require('util');
 const Hoek = require('@hapi/hoek');
+const Bourne = require('@hapi/bourne');
 
 class SchemaResolver {
-    constructor(root, { subSchemas, types, refineType, strictMode, extensions = [] }) {
+    constructor(root, { subSchemas, refineType, strictMode, extensions = [] }) {
```
Comment (tlivings) [inline lib/resolver.js:7]: Note to self: breaking change.

File: lib/resolver.js (hunk 1 above)
Comment (sprootshift) [reply lib/resolver.js:26]: Same as if the input is not string or does not validate with the regular expression. The coerce method returns `undefined`, which becomes the object or array value. This part of the code is actually adopted from Joi version 16 release notes: https://github.com/hapijs/joi/issues/2037

**Comment by sprootshift** [comment]
Thank you. Could you please release a new version? I will then update matching hapi-openapi pull request https://github.com/krakenjs/hapi-openapi/pull/176 to use npm enjoi release.

**Comment by tlivings** [comment]
Will be doing so tomorrow morning thanks.

**Comment by tlivings** [comment]
Published as 7.0

**Pull request merged**